### PR TITLE
Snapdeals unsealing fixes

### PR DESCRIPTION
**Pull request by magik6k**
## Related Issues
Depends on https://github.com/filecoin-project/go-state-types/pull/191
Depends on https://github.com/filecoin-project/filecoin-ffi/pull/402

(todo link relevant issues)

## Additional Info

Testing:
* [ ] ITests
  * [x] Existing itests pass
  * [ ] Test unseal on a worker with no long-term-storage, verify files cleaned up correctly
* [x] Manual testing

## Checklist

Before you mark the PR ready for review, please make sure that:

- [ ] Commits have a clear commit message.
- [ ] PR title is in the form of of `<PR type>: <area>: <change being made>`
  - example: ` fix: mempool: Introduce a cache for valid signatures`
  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
  - `area`, e.g. api, chain, state, market, mempool, multisig, networking, paych, proving, sealing, wallet, deps
- [ ] New features have usage guidelines and / or documentation updates in
  - [ ] [Lotus Documentation](https://lotus.filecoin.io)
  - [ ] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] Tests exist for new functionality or change in behavior
- [ ] CI is green

### CODE REVIEW

**Comment by Reiers** [comment]
Testing:
* [ ] ITests
  * [x] Existing itests pass
  * [x] Test unseal on a worker with no long-term-storage, verify files cleaned up correctly
* [x] Manual testing
  * [x] Test normal unseal
    * [x] Works on a worker with only sealing paths
    * [x] Unsealed file is moved to long-term storage after unsealing
    * [x] No files are left in sealing paths after unseal is done
    * [x] Sealed + Unsealed files are in long-term-storage
  * [x] Test snapdeals unseal
    * [x] Without sector-key in storage
      * [x] Works on a worker with only sealing paths
      * [x] Unsealed file is moved to long-term storage after unsealing
      * [x] No files are left in sealing paths after unseal is done
      * [x] Update + Unsealed files are in long-term-storage
    * [x] With sector-key in storage already
      * [x] Works on a worker with only sealing paths
      * [x] Unsealed file is moved to long-term storage after unsealing
      * [x] No files are left in sealing paths after unseal is done
      * [x] Update + Unsealed files are in long-term-storage

Regular deal sectors unsealing on a worker works - it moves the unsealed sectors to the long term storage, deletes the leftover files on the worker. So far great! 

BUT unsealing Snap Deals, still has issues unfortunately, it does everything right apart from the task itself + lotus-miner sectors unseal hangs until the task is complete. 

My journal from testing below:


Before UNS of snap deal:

`lotus-miner storage find 21139`
```
In FILECOIN_HYPER_SERVER (Sealed, Cache)
	Sealing: false; Storage: true
	Local (/nfs/TechHedgeUnit3)
In FILECOIN_ULTRA_SERVER (Update, UpdateCache)
	Sealing: false; Storage: true
	Local (/nfs/TechHedgeUnit2)
In TECHHEDGE5_STORAGE_SERVER (Unsealed)
	Sealing: false; Storage: true
	Local (/nfs/TechHedgeUnit5)
```

`lotus-miner storage redeclare --all
`
```
lotus-miner storage find 21139
In FILECOIN_HYPER_SERVER (Sealed, Cache)
	Sealing: false; Storage: true
	Local (/nfs/TechHedgeUnit3)
In FILECOIN_ULTRA_SERVER (Update, UpdateCache)
	Sealing: false; Storage: true
	Local (/nfs/TechHedgeUnit2)
```

`lotus-miner sectors unseal 21139` <- **the cmd hangs until the task is which is annoying (if the task takes around 3 hours)**

**Result:**

- UNS 2 minutes
- deletes files on worker
- sends it to long-term-storage.

**After a 2 min UNS:**
```
lotus-miner storage find 21139
In FILECOIN_HYPER_SERVER (Sealed, Cache)
	Sealing: false; Storage: true
	Local (/nfs/TechHedgeUnit3)
In FILECOIN_ULTRA_SERVER (Update, UpdateCache)
	Sealing: false; Storage: true
	Local (/nfs/TechHedgeUnit2)
In TECHHEDGE4_STORAGE_SERVER (Unsealed)
	Sealing: false; Storage: true
	Local (/nfs/TechHedgeUnit4)
```
Unsealed Size, my guess it’s broken due to short UNS time: 
`34359738378 aug.   3 12:22 s-t022352-21139`

Let me know what kind of logs you need @magik6k , if you don't see anything obvious right away.

**Comment by magik6k** [comment]
The go-state-types PR we depend on here got merged after nv21 changes, which break some integration tests without synth-porep ffi changes

Marking as blocked for now, those failing tests should resolve themselves after rebasing on synth-porep code.

**Comment by rjan90** [comment]
@magik6k Can we turn this back to unblocked now that [nv21 skeleton has been merged ](https://github.com/filecoin-project/lotus/pull/11201)into master?

File: extern/filecoin-ffi
Comment (shrenujbansal) [inline extern/filecoin-ffi:1]: Is this supposed to be here?

File: storage/sealer/ffiwrapper/sealer_cgo.go [hunk 1]
```diff
@@ -403,91 +404,188 @@ func (sb *Sealer) pieceCid(spt abi.RegisteredSealProof, in []byte) (cid.Cid, err
 	return pieceCID, werr()
 }
 
-func (sb *Sealer) tryDecodeUpdatedReplica(ctx context.Context, sector storiface.SectorRef, commD cid.Cid, unsealedPath string, randomness abi.SealRandomness) (bool, error) {
-	replicaPath, done, err := sb.sectors.AcquireSector(ctx, sector, storiface.FTUpdate|storiface.FTUpdateCache, storiface.FTNone, storiface.PathSealing)
+func (sb *Sealer) maybeAcquireUpdatePath(ctx context.Context, sector storiface.SectorRef) (string, func(), error) {
```
Comment (shrenujbansal) [inline storage/sealer/ffiwrapper/sealer_cgo.go:407]: I would suggest just calling this acquireUpdatePath and then documenting the cases where you will or will not receive the path in the code

File: storage/sealer/ffiwrapper/sealer_cgo.go [hunk 2]
```diff
@@ -403,91 +404,188 @@ func (sb *Sealer) pieceCid(spt abi.RegisteredSealProof, in []byte) (cid.Cid, err
...
+	// copy so that the sector doesn't get removed from a long-term storage path
+	replicaPath, done, err := sb.sectors.AcquireSectorCopy(ctx, sector, storiface.FTUpdate, storiface.FTNone, storiface.PathSealing)
```
Comment (shrenujbansal) [inline storage/sealer/ffiwrapper/sealer_cgo.go:409]: why is the var holding a func() called "done"?
Maybe have a more descriptive name. "done" sounds like a bool

File: storage/sealer/ffiwrapper/sealer_cgo.go [hunk 3]
```diff
@@ -403,91 +404,188 @@ func (sb *Sealer) pieceCid(spt abi.RegisteredSealProof, in []byte) (cid.Cid, err
...
+		return "", done, xerrors.Errorf("reading updated replica: %w", err)
 	}
-	defer done()
 
-	sealedPaths, done2, err := sb.AcquireSectorKeyOrRegenerate(ctx, sector, randomness)
+	return replicaPath.Update, done, nil
+}
+
+func (sb *Sealer) decodeUpdatedReplica(ctx context.Context, sector storiface.SectorRef, commD cid.Cid, updatePath, unsealedPath string, randomness abi.SealRandomness) error {
+	keyPaths, done2, err := sb.acquireSectorKeyOrRegenerate(ctx, sector, randomness)
```
Comment (shrenujbansal) [inline storage/sealer/ffiwrapper/sealer_cgo.go:420]: Same comment as above about the naming of func() return

File: storage/sealer/ffiwrapper/types.go [hunk 4]
```diff
@@ -11,6 +11,7 @@ type SectorProvider interface {
 	// * returns storiface.ErrSectorNotFound if a requested existing sector doesn't exist
 	// * returns an error when allocate is set, and existing isn't, and the sector exists
 	AcquireSector(ctx context.Context, id storiface.SectorRef, existing storiface.SectorFileType, allocate storiface.SectorFileType, ptype storiface.PathType) (storiface.SectorPaths, func(), error)
+	AcquireSectorCopy(ctx context.Context, id storiface.SectorRef, existing storiface.SectorFileType, allocate storiface.SectorFileType, ptype storiface.PathType) (storiface.SectorPaths, func(), error)
```
Comment (shrenujbansal) [inline storage/sealer/ffiwrapper/types.go:14]: add comment for this API please

File: storage/sealer/ffiwrapper/sealer_cgo.go [hunk 5]
```diff
@@ -403,91 +404,188 @@ func (sb *Sealer) pieceCid(spt abi.RegisteredSealProof, in []byte) (cid.Cid, err
...
+	}
+
+	return nil
 }
 
-func (sb *Sealer) AcquireSectorKeyOrRegenerate(ctx context.Context, sector storiface.SectorRef, randomness abi.SealRandomness) (storiface.SectorPaths, func(), error) {
-	paths, done, err := sb.sectors.AcquireSector(ctx, sector, storiface.FTSealed|storiface.FTCache, storiface.FTNone, storiface.PathStorage)
+func (sb *Sealer) acquireSectorKeyOrRegenerate(ctx context.Context, sector storiface.SectorRef, randomness abi.SealRandomness) (storiface.SectorPaths, func(), error) {
+	// copy so that the files aren't removed from long-term storage
+	paths, done, err := sb.sectors.AcquireSectorCopy(ctx, sector, storiface.FTSealed|storiface.FTCache, storiface.FTNone, storiface.PathSealing)
```
Comment (shrenujbansal) [inline storage/sealer/ffiwrapper/sealer_cgo.go:459]: mirroring above comment for func var name

File: storage/sealer/ffiwrapper/sealer_cgo.go [hunk 6]
```diff
@@ -403,91 +404,188 @@ func (sb *Sealer) pieceCid(spt abi.RegisteredSealProof, in []byte) (cid.Cid, err
...
 			return xerrors.Errorf("acquire unsealed sector path (allocate): %w", err)
 		}
-		defer done()
-
-		pf, err = partialfile.CreatePartialFile(maxPieceSize, unsealedPath.Unsealed)
-		if err != nil {
-			return xerrors.Errorf("create unsealed file: %w", err)
-		}
-
 	case err == nil:
```
Comment (shrenujbansal) [inline storage/sealer/ffiwrapper/sealer_cgo.go:572]: add a comment saying "noop" here?

**Review by rjan90 — APPROVED** [review]

**Review by jennijuju — APPROVED** [review]

**Pull request merged**